File: trady-backend/src/chart/domain/input.daily.ts
import { TimeSeriesType } from 'chart/model';
import { LoadChartInputBase } from './input.base';

export class LoadChartInputDaily implements LoadChartInputBase {
  timeSeriesType: Readonly<TimeSeriesType> = 'TIME_SERIES_DAILY';

  timeSeriesKey: Readonly<string> = 'Time Series (Daily)';

  symbol: Readonly<string>;

  constructor(symbol: string) {
    this.symbol = symbol;
  }

  toDescriptor(): Readonly<string> {
    return `${this.symbol}.${this.timeSeriesType}`;
  }

  toReferenceUrl(
    prefix: Readonly<String>,
    apiKey: Readonly<string>,
  ): Readonly<string> {
    return (
      `${prefix}` +
      `?function=${this.timeSeriesType}` +
      `&symbol=${this.symbol}` +
      `&apikey=${apiKey}`
    );
  }

  toTimeSeriesKey(): Readonly<string> {
    return this.timeSeriesKey;
  }
}

File: stock-tracker/src/chart_loader/input_daily.rs
use crate::chart_loader::{Creator, LoadChartInputTrait};

pub struct InputDaily {
  timeseries_type: String,
  timeseries_key: String,
  symbol: String,
}

impl Creator for InputDaily {
  fn new(symbol: String) -> Self {
    Self {
      timeseries_type: String::from("TIME_SERIES_DAILY"),
      timeseries_key: String::from("Time Series (Daily)"),
      symbol,
    }
  }
}

impl LoadChartInputTrait for InputDaily {
  fn to_descriptor(&self) -> String {
    format!("{}.{}", self.symbol, self.timeseries_type)
  }

  fn to_query_string(&self, prefix: String, apikey: String) -> String {
    format!(
      "{}?function={}&symbol={}&apikey={}",
      prefix, self.timeseries_type, self.symbol, apikey
    )
  }

  fn to_timeseries_key(&self) -> String {
    self.timeseries_key.clone()
  }
}
